### PR TITLE
Tbtc tests

### PR DESCRIPTION
**Pull request by michalsmiarowski**
Close: #439 

Adds tests for TBTC class in `threshold-ts` lib. Also resolves other tests that might have not worked because of some tbtc libs.

### CODE REVIEW

**Comment by mhluongo** [comment]
Can we get a PR description that might make this easier to review?

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 1]
```diff
@@ -0,0 +1,739 @@
...
+
+    beforeEach(async () => {
+      jest
+        .spyOn(mockBitcoinClient, "getTransactionConfirmations")
+        .mockImplementation(async () => mockTransactionConfirmations)
+      transactionConfirmationsResult = await tBTC.getTransactionConfirmations(
+        mockTransactionHash
+      )
+    })
+    test("should get revealed deposit", () => {
```
Comment (r-czajkowski) [inline src/threshold-ts/tbtc/__test__/tbtc.test.ts:579]: Same as above.

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 2]
```diff
@@ -0,0 +1,739 @@
...
+    })
+  })
+
+  describe("getTransactionConfirmations", () => {
+    let transactionConfirmationsResult: number
+    const mockTransactionConfirmations = 3
+    const mockTransactionHash = TransactionHash.from(
+      "9eb901fc68f0d9bcaf575f23783b7d30ac5dd8d95f3c83dceaa13dce17de816a"
+    )
+    console.log("mock t hash!: ", mockTransactionHash)
```
Comment (r-czajkowski) [inline src/threshold-ts/tbtc/__test__/tbtc.test.ts:569]: Do we need this console log?

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 3]
```diff
@@ -0,0 +1,739 @@
...
+        "_calculateOptimisticMintingAmountAndFee"
+      )
+
+      mockCalculateOptimisticMintingAmountAndFee.mockReturnValue({
+        amountToMint: mockAmountToMint,
+        optimisticMintFee: mockOptimisticMintFee,
+      })
+      estimatedFees = await tBTC.getEstimatedFees(mockDepositAmount)
+    })
+    test("should estimate fees", () => {
```
Comment (r-czajkowski) [inline src/threshold-ts/tbtc/__test__/tbtc.test.ts:500]: Same as above.

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+        await tBTC.findAllUnspentTransactionOutputs(mockDepositAddress)
+    })
+    test("should call a proper function with proper parameters", async () => {
+      expect(mockBitcoinClient.findAllUnspentTransactionOutputs).toBeCalledWith(
+        mockDepositAddress
+      )
+    })
+    test("should return utxos correctly", async () => {
+      expect(findAllUnspentTransactionOutputsResult).toEqual([testnetUTXO])
+    })
```
Comment (r-czajkowski) [inline src/threshold-ts/tbtc/__test__/tbtc.test.ts:464]: Let's add empty lines between `test` and `beforeEach` statements.

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 5]
```diff
@@ -0,0 +1,739 @@
...
+      })
+    })
+
+    describe("when provided btc address is not a PKH type address", () => {
+      beforeEach(async () => {
+        setUpdeposiScriptParameters()
+        setUpValidBTCAddress(true)
+        setUpIsPublicKeyHashTypeAddress(false)
+      })
+      test("should throw", async () => {
```
Comment (r-czajkowski) [inline src/threshold-ts/tbtc/__test__/tbtc.test.ts:408]: Same as above.

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 6]
```diff
@@ -0,0 +1,739 @@
...
+      })
+    })
+
+    describe("when wrong btc address is provided", () => {
+      beforeEach(async () => {
+        setUpdeposiScriptParameters()
+        setUpValidBTCAddress(false)
+        setUpIsPublicKeyHashTypeAddress(false)
+      })
+      test("should throw", async () => {
```
Comment (r-czajkowski) [inline src/threshold-ts/tbtc/__test__/tbtc.test.ts:388]: Let's add an empty line.

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 7]
```diff
@@ -0,0 +1,739 @@
...
+        expect(isPublicKeyHashTypeAddress).toHaveBeenCalledWith(
+          bitcoinAddressTestnet
+        )
+        expect(calculateDepositRefundLocktime).toHaveBeenCalledWith(
+          Math.floor(mockCurrentTime / 1000),
+          23328000
+        )
+        expect(getChainIdentifier).toBeCalledWith(ethAddress)
+      })
+      test("should create proper deposit script parameters", async () => {
```
Comment (r-czajkowski) [inline src/threshold-ts/tbtc/__test__/tbtc.test.ts:365]: Same as above.

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 8]
```diff
@@ -0,0 +1,739 @@
...
+
+  let tBTC: ITBTC
+  let tBTCMainnet: ITBTC
+
+  let bridge: EthereumBridge
+  let tbtcVault: Contract
+  let bitcoinClient: Client
+  let multicall: IMulticall
+  let bridgeContract: Contract
+  let token: Contract
```
Comment (r-czajkowski) [inline src/threshold-ts/tbtc/__test__/tbtc.test.ts:114]: Unused variables: `tBTCMainnet`,` tbtcVault`, `bitcoinClient`, `bridgeContract`, `token`.

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 5 above)
Comment (r-czajkowski) [inline src/threshold-ts/tbtc/__test__/tbtc.test.ts:408]: maybe should throw an error?

**Review by r-czajkowski — CHANGES_REQUESTED** [review]
Some tests fail because of `Cannot find module 'bcrypto.node'` error, even the tests in the `tbtc.test.ts` file. Can we fix it?

**Comment by mhluongo** [comment]
Are we planning to merge this, or should it be closed? Looks like it's danging right now

**Comment by michalsmiarowski** [comment]
> Are we planning to merge this, or should it be closed? Looks like it's danging right now

Yes. It did fall back a little since it wasn't that high of a priority, but I'm fixing it right now and we will try to push it out soon

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 9]
```diff
@@ -0,0 +1,739 @@
...
+          ethConfig.account
+        ),
+      })
+
+      expect(tBTC.vaultContract).toBe(mockTBTCVaultContract)
+      expect(tBTC.bridgeContract).toBe(mockBridgeContract)
+      expect(tBTC.tokenContract).toBe(mockTokenContract)
+      expect(tBTC.bitcoinNetwork).toBe(BitcoinNetwork.Testnet)
+    }
+    describe("with mocked bitcoin clinet", () => {
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 10]
```diff
@@ -0,0 +1,739 @@
...
+        network: BitcoinNetwork.Testnet,
+        credentials: [
+          {
+            host: "testhost",
+            port: 1,
+            protocol: "wss",
+          },
+        ],
+      }
+      beforeEach(() => {
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 11]
```diff
@@ -0,0 +1,739 @@
...
+      test("should create the real electrum client instance", () => {
+        expect(ElectrumClient).toBeCalled()
+      })
+    })
+
+    describe("with client not specified in the bitcoin config", () => {
+      const btcConfigWithNoClientSpecified: BitcoinConfig = {
+        network: BitcoinNetwork.Testnet,
+      }
+      beforeEach(() => {
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 12]
```diff
@@ -0,0 +1,739 @@
...
+
+  describe("suggestDepositWallet", () => {
+    let suggestDepositWalletResult: string | undefined
+    beforeEach(async () => {
+      suggestDepositWalletResult = await tBTC.suggestDepositWallet()
+    })
+    test("should call a proper function with proper parameters", async () => {
+      expect(bridge.activeWalletPublicKey).toBeCalledWith()
+    })
+    test("should suggest a desposit wallet correctly", async () => {
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 13]
```diff
@@ -0,0 +1,739 @@
...
+      ;(calculateDepositRefundLocktime as jest.Mock).mockImplementationOnce(
+        () => mockCalculateDepositRefundLocktimeResult
+      )
+      ;(computeHash160 as jest.Mock).mockImplementationOnce(
+        () => mockWalletPublicKeyHash
+      )
+      CryptoJS.lib.WordArray.random.mockImplementation(() => ({
+        toString: jest.fn().mockImplementation(() => mockBlindingFactor),
+      }))
+      jest
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts:342]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 14]
```diff
@@ -0,0 +1,739 @@
...
+      beforeEach(async () => {
+        setUpdeposiScriptParameters()
+        setUpValidBTCAddress(true)
+        setUpIsPublicKeyHashTypeAddress(true)
+        depositScriptParameters = await tBTC.createDepositScriptParameters(
+          ethAddress,
+          bitcoinAddressTestnet
+        )
+      })
+      test("should call proper functions with proper arguments", async () => {
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 1 above)
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 15]
```diff
@@ -0,0 +1,739 @@
...
+      treasuryFee: BigNumber.from(5000),
+    }
+
+    beforeEach(async () => {
+      ;(getRevealedDeposit as jest.Mock).mockImplementationOnce(
+        () => mockRevealedDeposit
+      )
+      revealedDepositResult = await tBTC.getRevealedDeposit(testnetUTXO)
+    })
+    test("should get revealed deposit", () => {
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 16]
```diff
@@ -0,0 +1,739 @@
...
+    beforeEach(async () => {
+      ;(revealDeposit as jest.Mock).mockImplementationOnce(
+        () => mockRevealedDepositTxHash
+      )
+      revealDepositResult = await tBTC.revealDeposit(
+        testnetUTXO,
+        mockDepositScriptParameters
+      )
+    })
+    test("should reveal deposit properly", () => {
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 3 above)
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 4 above)
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 17]
```diff
@@ -0,0 +1,739 @@
...
+    test("should call proper function with proper parameters", () => {
+      expect(calculateDepositAddress).toHaveBeenCalledWith(
+        mockDepositScriptParameters,
+        btcConfig.network,
+        true
+      )
+    })
+    test("should calculate deposit address correctly", () => {
+      expect(calculateDepositAddressResult).toBe(mockDepositAddress)
+    })
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 5 above)
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 6 above)
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 7 above)
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 3e72a34d95575a7bd5e753700917f29cfa27832d

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 8 above)
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: a4a6ff1eefe5b6aa081aef63628fc2fc7b4ad88a

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 7372eb86e267322da3686c50496aba647396c7ba

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 5 above)
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 7372eb86e267322da3686c50496aba647396c7ba

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 18]
```diff
@@ -0,0 +1,739 @@
...
+      const amountInSatoshi = 1 * 10 ** 8 // 1 BTC
+      const minimumNumberOfConfirmationsResult =
+        tBTC.minimumNumberOfConfirmationsNeeded(amountInSatoshi)
+      const amountInSatoshi2 = 9 * 10 ** 8 // 9 BTC
+      const minimumNumberOfConfirmationsResult2 =
+        tBTC.minimumNumberOfConfirmationsNeeded(amountInSatoshi2)
+      expect(minimumNumberOfConfirmationsResult2).toBe(6)
+      expect(minimumNumberOfConfirmationsResult2).toBe(6)
+    })
+  })
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: b84b9a22357583146e5e131ff13eb19e1a2a66b0

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 19]
```diff
@@ -0,0 +1,739 @@
...
+        mockTransactionHash.toString(),
+        mockDepositOutputIndex,
+        "big-endian"
+      )
+      expect(EthereumBridge.buildDepositKey).toHaveBeenCalledWith(
+        mockTransactionHash,
+        mockDepositOutputIndex
+      )
+      expect(buildDepositKeyResult).toBe(mockDepositKey)
+    })
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: b84b9a22357583146e5e131ff13eb19e1a2a66b0

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 20]
```diff
@@ -0,0 +1,739 @@
...
+      mockFindAllCancelledDepositsFunction = jest.spyOn(
+        tBTC as any,
+        "_findAllCancelledDeposits"
+      )
+      mockFindAllCancelledDepositsFunction.mockResolvedValue([])
+
+      result = await tBTC.bridgeTxHistory(mockDepositor)
+    })
+
+    test("should create the bridge history properly", () => {
```
[the remark this answers is not in the document]
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: d3ac52443a61fca59f4aee1f1118c41d9e58ba1e

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 2 above)
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: f8fa81307ced055059c5035980d5e9a46114c02d

File: src/posthog/index.ts [hunk 21]
```diff
@@ -4,8 +4,11 @@ import { getEnvVariable } from "../utils/getEnvVariable"
 import { PosthogEvent } from "../types/posthog"
 
 export const init = () => {
-  const apiKey = getEnvVariable(EnvVariable.POSTHOG_API_KEY)
-  const apiHost = getEnvVariable(EnvVariable.POSTHOG_HOSTNAME_HTTP)
+  // const apiKey = getEnvVariable(EnvVariable.POSTHOG_API_KEY)
+  // const apiHost = getEnvVariable(EnvVariable.POSTHOG_HOSTNAME_HTTP)
+
+  const apiKey = "phc_FANjzFhmW7Aj1UTklEuErRRAfrK4JkXN63bvJIaZ4jl"
+  const apiHost = "https://app.posthog.com"
```
Comment (r-czajkowski) [inline src/posthog/index.ts:11]: I believe this is an unnecessary change.

File: src/posthog/index.ts [hunk 22]
```diff
@@ -31,6 +34,7 @@ export const identify = (ethAddress: string) => {
 // you must make a `pageview` call manually.
 export const capturePageview = () => {
   posthog.capture("$pageview")
+  console.log("page view captured!")
```
Comment (r-czajkowski) [inline src/posthog/index.ts:37]: Same as above. Unnecessary console log.

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 23]
```diff
@@ -0,0 +1,781 @@
...
+        "9eb901fc68f0d9bcaf575f23783b7d30ac5dd8d95f3c83dceaa13dce17de816a"
+      ),
+  },
+  computeHash160: jest.fn(),
+}))
+
+jest.mock("@keep-network/tbtc-v2.ts/dist/src", () => ({
+  EthereumBridge: jest.fn(),
+  ElectrumClient: jest.fn(),
+}))
```
Comment (r-czajkowski) [inline src/threshold-ts/tbtc/__test__/tbtc.test.ts:110]: We have the same code in `setupTests.js`. Do we need this here?

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts [hunk 24]
```diff
@@ -0,0 +1,781 @@
...
+  abi: [],
+}))
+
+jest.mock("@keep-network/tbtc-v2.ts/dist/src/deposit", () => ({
+  calculateDepositAddress: jest.fn(),
+  calculateDepositRefundLocktime: jest.fn(),
+  DepositScriptParameters: jest.fn(),
+  revealDeposit: jest.fn(),
+  getRevealedDeposit: jest.fn(),
+}))
```
Comment (r-czajkowski) [inline src/threshold-ts/tbtc/__test__/tbtc.test.ts:66]: Same as above.

**Review by r-czajkowski — CHANGES_REQUESTED** [review]

**Review by r-czajkowski — CHANGES_REQUESTED** [review]
Let's fix `useFetchTvl` tests. The issue here is that the `Multicall` class from `threshold-ts` uses the zero address for chain id `1337`(chain id fetched from `.env` file) because the `REACT_APP_MULTICALL_ADDRESS` env is missed and then the `getContract` function throws an error if the address is zero. See:
- https://github.com/threshold-network/token-dashboard/blob/main/src/threshold-ts/multicall/index.ts#L51-L60
- https://github.com/threshold-network/token-dashboard/blob/main/src/threshold-ts/utils/contract.ts#L27-L29

So we can fix test in 2 ways:
- mock the `supportedChainId` variable(because this variable is passed to the `Threshold` class constructor.
   `useFetchTvl.test.tsx`
   ```ts
   jest.mock("../../utils/getEnvVariable", () => ({
  ...(jest.requireActual("../../utils/getEnvVariable") as {}),
  supportedChainId: 1,
   }))
   ```
- or we can create a new file `.env.test` where we define the mocked address for the `Multicall` contract- so if we run test `yarn test` the env variables from `.env.test` file will be fetched.
   ```
   REACT_APP_SUPPORTED_CHAIN_ID=1337
   REACT_APP_MULTICALL_ADDRESS=0x086813525A7dC7dafFf015Cdf03896Fd276eab60
   ```

**Comment by michalsmiarowski** [comment]
> Let's fix `useFetchTvl` tests. The issue here is that the `Multicall` class from `threshold-ts` uses the zero address for chain id `1337`(chain id fetched from `.env` file) because the `REACT_APP_MULTICALL_ADDRESS` env is missed and then the `getContract` function throws an error if the address is zero. See:
> 
> * https://github.com/threshold-network/token-dashboard/blob/main/src/threshold-ts/multicall/index.ts#L51-L60
> * https://github.com/threshold-network/token-dashboard/blob/main/src/threshold-ts/utils/contract.ts#L27-L29
> 
> So we can fix test in 2 ways:
> 
> * mock the `supportedChainId` variable(because this variable is passed to the `Threshold` class constructor.
>   `useFetchTvl.test.tsx`
>   ```ts
>   jest.mock("../../utils/getEnvVariable", () => ({
>   ...(jest.requireActual("../../utils/getEnvVariable") as {}),
>   supportedChainId: 1,
>   }))
>   ```
> * or we can create a new file `.env.test` where we define the mocked address for the `Multicall` contract- so if we run test `yarn test` the env variables from `.env.test` file will be fetched.
>   ```
>   REACT_APP_SUPPORTED_CHAIN_ID=1337
>   REACT_APP_MULTICALL_ADDRESS=0x086813525A7dC7dafFf015Cdf03896Fd276eab60
>   ```

That's awesome! Thanks for figuring that out because I wasn't sure why this test doesn't passing through. I've choosen the second solution as it looks like more global one so it will be also helpful for the future tests in which we will use `Multicall` class from `threshold-ts` lib. You can see the fix here: 6c04e845afac8f55a63eaeabf58606a2c881e80a

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 23 above)
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 86b950b592e6adba04532133c7d808efe038f87e

File: src/threshold-ts/tbtc/__test__/tbtc.test.ts (hunk 24 above)
Comment (michalsmiarowski) [reply src/threshold-ts/tbtc/__test__/tbtc.test.ts]: 86b950b592e6adba04532133c7d808efe038f87e

**Review by r-czajkowski — APPROVED** [review]
LGTM!

**Pull request merged**